src/cpp/server/server_context.cc

```diff
@@ -219,7 +219,17 @@ bool ServerContext::CompletionOp::FinalizeResult(void** tag, bool* status) {
   // interceptors now
   lock.Unlock();
 
-  if (call_cancel && reactor_ != nullptr) {
+  // The cancellation state atomic variable needs acquire order on fail since
+  // the results generated in OnStarted need to be visible before the current
+  // invocation to OnCancel. The fail case does not need any release ordering
+  // since no results have been produced in this function except those which
+  // were controlled under the cancellation lock that is also used by
+  // IsCancelled. In the success case, there is no need for ordering since no
+  // results have been produced.
+  if (call_cancel && reactor_ != nullptr &&
+      !reactor_->MaybeChangeCancellationState(
+          internal::ServerReactor::CancellationState::ON_CANCEL_NEEDED,
+          std::memory_order_relaxed, std::memory_order_acquire)) {
```

Comment:
Looks like gcc 4.9 doesn't allow the failure case to be stronger than the success case (which, I'd admit is odd looking in any case) so I'll have to strengthen the success case here.

Reply:
Added a comment and TODO in the code to explain this issue.

Reply:
Oh, yes, I believe that's a requirement in the C++11 which was removed in C++17.